### PR TITLE
initial implementation, UTs, and SDD for BufferRepeater

### PR DESCRIPTION
**Pull request by LeStarch**
| | |
|:---|:---|
|**_Originating Project/Creator_**| |
|**_Affected Component_**|  |
|**_Affected Architectures(s)_**|  |
|**_Related Issue(s)_**|  |
|**_Has Unit Tests (y/n)_**|  y |
|**_Builds Without Errors (y/n)_**| y  |
|**_Unit Tests Pass (y/n)_**|  y |
|**_Documentation Included (y/n)_**|  y |

---
## Change Description

Adds a buffer repeater component. This component receives an incoming `Fw.BufferSend` an repeats it to a set of output ports. Data is properly copied and then deallocated.

## Rationale

Sometimes buffer port calls need to be clones in the style of `Svc.ComSplitter`.

## Testing/Review Recommendations

Run UTs or observe CI doing that.

## Future Work

None.

### CODE REVIEW

**Comment by LeStarch** [comment]
@thomas-bc this adds a buffer-repeater.  This component repeats an incoming Fw.Buffer port to a set of outgoing Fw.Buffer reports.  It ensures new allocations for each.

Would you review this?  I can have @timcanham look too, if you'd like.

File: Svc/BufferRepeater/docs/sdd.md [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+## 3. Design
+
+The buffer repeater has a single `portIn` port of type `Fw.BufferSend` and an array of `portOut` ports of the same type.
+Each port call received on `portIn` is repeated to each connected `portOut` port. These repetitions emit a copy of the
+`Fw::Buffer` data.
+
+## 3.1 Ports
+
+`Svc::ComQueue` has the following ports:
```
Comment (thomas-bc) [inline Svc/BufferRepeater/docs/sdd.md:28]: Should be `Svc::BufferRepeater`

**Review by thomas-bc — APPROVED** [review]
Looks good to me! There's a small copy/paste remnant in the SDD that should be corrected -- but I'll go ahead and approve this.

File: Svc/BufferRepeater/BufferRepeater.cpp [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    BufferRepeaterComponentBase::init(instance);
+}
+
+BufferRepeater ::~BufferRepeater() {}
+
+void BufferRepeater ::configure(BufferRepeater::BufferRepeaterFailureOption allocation_failure_response) {
+    this->m_allocation_failure_response = allocation_failure_response;
+}
+
+bool BufferRepeater ::check_allocation(FwIndexType index,
```
Comment (timcanham) [inline Svc/BufferRepeater/BufferRepeater.cpp:36]: Should we assert that index < number of ports?

File: Svc/BufferRepeater/BufferRepeater.cpp [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+void BufferRepeater ::portIn_handler(NATIVE_INT_TYPE portNum, /*!< The port number*/
+                                     Fw::Buffer& buffer       /*!< The serialization buffer*/
+) {
+    FW_ASSERT(this->m_allocation_failure_response < NUM_BUFFER_REPEATER_FAILURE_OPTIONS);
+    for (FwIndexType i = 0; i < NUM_PORTOUT_OUTPUT_PORTS; i++) {
+        if (isConnected_portOut_OutputPort(i)) {
+            Fw::Buffer new_allocation = this->allocate_out(0, buffer.getSize());
+            if (this->check_allocation(i, new_allocation, buffer)) {
+                // Clone the data and send it
+                ::memcpy(new_allocation.getData(), buffer.getData(), buffer.getSize());
```
Comment (timcanham) [inline Svc/BufferRepeater/BufferRepeater.cpp:74]: Don't we have to set the destination buffer size to be the same as the source so users know how much data there is?

**Review by timcanham — CHANGES_REQUESTED** [review]

File: Svc/BufferRepeater/BufferRepeater.cpp (hunk 3 above)
Comment (LeStarch) [reply Svc/BufferRepeater/BufferRepeater.cpp:74]: Yes, yes we do!  Good catch.  Also need to put this in UTs.

File: Svc/BufferRepeater/BufferRepeater.cpp (hunk 2 above)
Comment (LeStarch) [reply Svc/BufferRepeater/BufferRepeater.cpp:36]: yup

**Comment by LeStarch** [comment]
@thomas-bc @timcanham want to check the fixes?

**Review by timcanham — APPROVED** [review]

**Pull request merged**